### ISSUE TITLE
Cloning repo produces "zeroPaddedFilemode: contains zero-padded file modes" error

### ISSUE THREAD

**Issue by tfar** [michaeljones/breathe#340]
I just tried to clone the breathe repo, however git fails with an error.

```
 ~/Development> git clone https://github.com/michaeljones/breathe
Cloning into 'breathe'...
remote: Counting objects: 6089, done.
remote: Compressing objects: 100% (6/6), done.
error: object 28e789bea886bcba4caccb8f6b584aee73bdbb4a: zeroPaddedFilemode: contains zero-padded file modes
fatal: Error in object
fatal: index-pack failed
 ~/Development> git --version
git version 2.13.2
```

**Comment by t-b** [comment]
@tfar Works here. Can you try again? Does other repos on github work?

**Comment by tfar** [comment]
Still failing:
```
 ~> git clone https://github.com/michaeljones/breathe
Cloning into 'breathe'...
remote: Counting objects: 6089, done.
error: object 28e789bea886bcba4caccb8f6b584aee73bdbb4a: zeroPaddedFilemode: contains zero-padded file modes
fatal: Error in object
fatal: index-pack failed
```

**Comment by t-b** [comment]
@tfar Seems the repository has some invalid filemode data in it. See e.g. http://git.661346.n2.nabble.com/Zero-padded-file-modes-td7595485.html.

Does `git -c transfer.fsckObjects=false clone https://github.com/michaeljones/breathe` work on your end?

**Comment by tfar** [comment]
@t-b That also fails here.

```
 ~> cd /tmp/
 /p/tmp> git -c transfer.fsckObjects=false clone https://github.com/michaeljones/breathe
Cloning into 'breathe'...
remote: Counting objects: 6089, done.
error: object 28e789bea886bcba4caccb8f6b584aee73bdbb4a: zeroPaddedFilemode: contains zero-padded file modes
fatal: Error in object
fatal: index-pack failed
```

**Comment by melvinvermeeren** [comment]
For future reference: use

`git clone --config transfer.fsckobjects=false --config receive.fsckobjects=false --config fetch.fsckobjects=false https://github.com/michaeljones/breathe`

This should work no matter how fsck is enabled on your configuration.

**Comment by melvinvermeeren** [comment]
Hi. After further thinking I come to the conclusion that this is quite bad. `fsck` is an useful feature and I believe it is getting used more and more by both clients and servers.

Unfortunately fixing this will probably require rewriting a massive part of project's history and force pushing. I will look into it.

**Comment by melvinvermeeren** [comment]
# Rewriting the history

```
$ mv breathe breathe-old-prerewrite
$ mkdir breathe; cd breathe; git init; cd ..
$ cd breathe-old-prerewrite
$ git fast-export --signed-tags=warn-strip --all | (cd ../breathe && git fast-import)
warning: Stripping signature from tag f250c0c77fdbfa11e9bf8051f93c1d959aa7da8a
warning: Stripping signature from tag 709c07832787fe624772960b9f57a0cced830cd9
warning: Stripping signature from tag 5b4583307416541bf43f5d55cb203633d0a76596
/usr/lib/git-core/git-fast-import statistics:
---------------------------------------------------------------------
Alloc'd objects:      10000
Total objects:         6306 (       873 duplicates                  )
      blobs  :         1993 (         0 duplicates       1024 deltas of       1985 attempts)
      trees  :         3271 (       873 duplicates       3141 deltas of       3232 attempts)
      commits:         1036 (         0 duplicates          0 deltas of          0 attempts)
      tags   :            6 (         0 duplicates          0 deltas of          0 attempts)
Total branches:          38 (        37 loads     )
      marks:        1048576 (      3029 unique    )
      atoms:            258
Memory total:          2579 KiB
       pools:          2110 KiB
     objects:           468 KiB
---------------------------------------------------------------------
pack_report: getpagesize()            =       4096
pack_report: core.packedGitWindowSize = 1073741824
pack_report: core.packedGitLimit      = 35184372088832
pack_report: pack_used_ctr            =       1776
pack_report: pack_mmap_calls          =        399
pack_report: pack_open_windows        =          1 /          1
pack_report: pack_mapped              =    5335617 /    5335617
---------------------------------------------------------------------
```

Backup for old hashes: https://github.com/melvinvermeeren/breathe-before-rewrite-issue-340-backup

# Signing tags

Three tags exists that were PGP signed (4.8.0, 4.9.0, 4.9.1). I resigned them for the new hashes.

# Commit signatures

All PGP commit signatures have been lost by rewriting the history. The exception is the new HEAD which is resigned by running `git commit --amend`.

**Comment by melvinvermeeren** [comment]
All branches and tags have been checked and appear ok. A fresh clone does not error now.

Closing as it should be fixed for good now.

**Comment by michaeljones** [comment]
I was pretty shocked by this one to be honest, though I can see the reasoning! 

I have compared the new & old repositories at the equivalent of this commit: d5b6ab614bb14d6316cf69f4304921828ea7b271 and found them to be the same and all changes since then seem fine.

Thanks for the hard work @melvinvermeeren. Sorry that file mode corruption got in there somehow!

**Comment by melvinvermeeren** [comment]
@michaeljones

Yeah, this is indeed the big thing to avoid when working with git repositories. There was unfortunately no way around this.

I had made both a local backup and a remote backup (https://github.com/melvinvermeeren/breathe-before-rewrite-issue-340-backup) prior to doing these destructive operations.

What is omitted from the above posts is checking the current HEAD (4.9.1), which perhaps you did the same way as me. I did it by copying the pre-rewrite and post-rewrite repositories, deleting their `.git` folders and running `git diff --no-index breathe-old-prewrite breathe`. The conclusion was the same as yours, no differences.

The odd thing was GitHub said 1010 commits pre-rewrite and 1009 commits post-rewrite. Most likely this was the corrupt object causing the issue in the first place.
